File: src/ViewModel/PlayerViewModel.cpp
#include "PlayerViewModel.h"
#include "Player.h"
#include "Character.h"

PlayerViewModel::PlayerViewModel(Player* player, QObject* parent)
    : QObject(parent)
    , m_player(player)
{
    if (!m_player) return;

    // 直接连接 Model 的 Qt 信号 → 转发为本 VM 的信号
    connect(m_player, &Player::hpChanged, this, &PlayerViewModel::hpChanged);
    connect(m_player, &Player::maxHpChanged, this, &PlayerViewModel::maxHpChanged);
    connect(m_player, &Player::dying, this, &PlayerViewModel::dying);
    connect(m_player, &Player::died, this, &PlayerViewModel::died);
    connect(m_player, &Player::revived, this, &PlayerViewModel::revived);
    connect(m_player, &Player::handCardAdded, this, &PlayerViewModel::handCardAdded);
    connect(m_player, &Player::handCardRemoved, this, &PlayerViewModel::handCardRemoved);
    connect(m_player, &Player::handCardsChanged, this, &PlayerViewModel::handCardsChanged);
    connect(m_player, &Player::stateChanged, this, &PlayerViewModel::stateChanged);
}

PlayerViewModel::~PlayerViewModel()
{
    // Qt 信号槽在 QObject 析构时自动断开，无需手动管理
}

int PlayerViewModel::playerId() const { return m_player ? m_player->playerId() : -1; }
QString PlayerViewModel::displayName() const { return m_player ? m_player->displayName() : QString(); }
QString PlayerViewModel::characterName() const { return m_player ? m_player->characterName() : QString(); }

QString PlayerViewModel::skillName() const
{
    if (m_player && m_player->character()) {
        return QString::fromStdString(m_player->character()->skillName());
    }
    return QString();
}

QString PlayerViewModel::skillDescription() const
{
    if (m_player && m_player->character()) {
        return QString::fromStdString(m_player->character()->skillDescription());
    }
    return QString();
}

int PlayerViewModel::hp() const { return m_player ? m_player->hp() : 0; }
int PlayerViewModel::maxHp() const { return m_player ? m_player->maxHp() : 0; }
double PlayerViewModel::hpRatio() const { return m_player ? m_player->hpRatio() : 0.0; }
bool PlayerViewModel::isAlive() const { return m_player ? m_player->isAlive() : false; }
bool PlayerViewModel::isDying() const { return m_player ? m_player->isDying() : false; }
bool PlayerViewModel::isFullHp() const { return m_player ? m_player->isFullHp() : false; }
int PlayerViewModel::handCardCount() const { return m_player ? m_player->handCardCount() : 0; }
bool PlayerViewModel::hasHandCards() const { return m_player ? m_player->hasHandCards() : false; }
int PlayerViewModel::handCardLimit() const { return m_player ? m_player->handCardLimit() : 0; }
bool PlayerViewModel::hasUsedKillThisTurn() const { return m_player ? m_player->hasUsedKillThisTurn() : false; }
bool PlayerViewModel::isWineEnhanced() const { return m_player ? m_player->isWineEnhanced() : false; }

int PlayerViewModel::equipCount() const
{
    return m_player ? static_cast<int>(m_player->equipCards().size()) : 0;
}

int PlayerViewModel::judgmentCount() const
{
    return m_player ? static_cast<int>(m_player->judgmentCards().size()) : 0;
}

Player* PlayerViewModel::player() const { return m_player; }

File: Unity/Scripts/ViewModel/PlayerViewModel.cs
using System.Collections.Generic;

/// <summary>
/// 玩家 ViewModel — 包装 Player 并转发其事件，方便 View 层只监听 ViewModel 而不直接依赖 Model
/// </summary>
public class PlayerViewModel
{
    private readonly Player _player;

    // Model 事件连接 ID（用于析构时断开）
    private readonly Dictionary<string, int> _connIds = new();

    public PlayerViewModel(Player player)
    {
        _player = player;
        ConnectModelEvents();
    }

    /// <summary>断开所有 Model 事件</summary>
    public void Dispose()
    {
        if (_player == null) return;

        _player.HpChanged.Disconnect(_connIds.GetValueOrDefault("hp", 0));
        _player.MaxHpChanged.Disconnect(_connIds.GetValueOrDefault("maxHp", 0));
        _player.Dying.Disconnect(_connIds.GetValueOrDefault("dying", 0));
        _player.Died.Disconnect(_connIds.GetValueOrDefault("died", 0));
        _player.Revived.Disconnect(_connIds.GetValueOrDefault("revived", 0));
        _player.HandCardAdded.Disconnect(_connIds.GetValueOrDefault("cardAdd", 0));
        _player.HandCardRemoved.Disconnect(_connIds.GetValueOrDefault("cardRem", 0));
        _player.HandCardsChanged.Disconnect(_connIds.GetValueOrDefault("cardsCh", 0));
        _player.StateChanged.Disconnect(_connIds.GetValueOrDefault("state", 0));

        _connIds.Clear();
    }

    // ==================== 身份标识 ====================

    public int PlayerId => _player?.PlayerId ?? -1;
    public string DisplayName => _player?.DisplayName ?? "";
    public string CharacterName => _player?.CharacterName ?? "";
    public string SkillName => _player?.Character?.SkillName ?? "";
    public string SkillDescription => _player?.Character?.SkillDescription ?? "";

    // ==================== 体力 ====================

    public int Hp => _player?.Hp ?? 0;
    public int MaxHp => _player?.MaxHp ?? 0;
    public float HpRatio => _player?.HpRatio ?? 0f;
    public bool IsAlive => _player?.IsAlive() ?? false;
    public bool IsDying => _player?.IsDying() ?? false;
    public bool IsFullHp => _player?.IsFullHp() ?? false;

    // ==================== 手牌 ====================

    public int HandCardCount => _player?.HandCardCount ?? 0;
    public List<Card> HandCards => _player?.HandCards ?? new List<Card>();
    public bool HasHandCards => _player?.HasHandCards() ?? false;
    public int HandCardLimit => _player?.HandCardLimit ?? 0;

    // ==================== 回合状态 ====================

    public bool HasUsedKillThisTurn => _player?.HasUsedKillThisTurn() ?? false;
    public bool IsWineEnhanced => _player?.IsWineEnhanced() ?? false;

    // ==================== 装备/判定区（预留）====================

    public int EquipCount => _player?.EquipCards.Count ?? 0;
    public int JudgmentCount => _player?.JudgmentCards.Count ?? 0;

    // ==================== 原始指针访问 ====================

    public Player Player => _player;

    // ==================== 事件（从 Model 转发）====================

    public readonly Signal<int> HpChanged = new();
    public readonly Signal<int> MaxHpChanged = new();
    public readonly Signal<Player> Dying = new();
    public readonly Signal<Player> Died = new();
    public readonly Signal<Player> Revived = new();
    public readonly Signal<Card> HandCardAdded = new();
    public readonly Signal<Card> HandCardRemoved = new();
    public readonly Signal HandCardsChanged = new();
    public readonly Signal StateChanged = new();

    // ==================== 事件转发 ====================

    private void ConnectModelEvents()
    {
        if (_player == null) return;

        _connIds["hp"] = _player.HpChanged.Connect(hp => HpChanged.Emit(hp));
        _connIds["maxHp"] = _player.MaxHpChanged.Connect(mh => MaxHpChanged.Emit(mh));
        _connIds["dying"] = _player.Dying.Connect(p => Dying.Emit(p));
        _connIds["died"] = _player.Died.Connect(p => Died.Emit(p));
        _connIds["revived"] = _player.Revived.Connect(p => Revived.Emit(p));
        _connIds["cardAdd"] = _player.HandCardAdded.Connect(c => HandCardAdded.Emit(c));
        _connIds["cardRem"] = _player.HandCardRemoved.Connect(c => HandCardRemoved.Emit(c));
        _connIds["cardsCh"] = _player.HandCardsChanged.Connect(() => HandCardsChanged.Emit());
        _connIds["state"] = _player.StateChanged.Connect(() => StateChanged.Emit());
    }
}
